File: eventvault/science/guardrail.py
"""
Science Guardrail for EventVault-R.

Implements the deletion condition from Section VI of the proposal:

    |F̂_D - F_ref| / |F_ref| ≤ ε_F  AND  ‖x̂_{c,D} - x_{c,ref}‖₂ ≤ ε_x
    (Eq. 13)

The guardrail evaluates residual layers in reverse detail order
(Section VI.C): if removing H3 is safe but removing H2 violates
the error bound, H3 may be purged while H2 remains locked.

This gives the controller fine-grained control rather than a
binary frame-level deletion decision.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Tuple

import numpy as np
from numpy.typing import NDArray

from eventvault.codec.dwt import WaveletDecomposition, reconstruct
from eventvault.science.source_extraction import (
    SourceMeasurement,
    extract_sources,
    measure_source,
)


@dataclass
class GuardrailResult:
    """
    Result of the Science Guardrail evaluation for a single layer.

    Attributes
    ----------
    layer : int
        Layer index evaluated (1-indexed detail level).
    is_safe : bool
        True if removing this layer maintains science bounds.
    max_flux_error : float
        Maximum relative flux error across all sources.
    max_centroid_error : float
        Maximum centroid displacement across all sources.
    num_sources_evaluated : int
        Number of sources used in the evaluation.
    violations : list
        Details of sources that violated bounds.
    """
    layer: int
    is_safe: bool
    max_flux_error: float
    max_centroid_error: float
    num_sources_evaluated: int
    violations: List[dict]


@dataclass
class FrameGuardrailDecision:
    """
    Complete guardrail decision for one frame.

    Attributes
    ----------
    frame_id : int
        Frame identifier.
    layer_decisions : dict
        {layer_idx: GuardrailResult} for each evaluated layer.
    purgeable_layers : list
        Layer indices that can be safely purged.
    locked_layers : list
        Layer indices that must be retained.
    """
    frame_id: int
    layer_decisions: Dict[int, GuardrailResult]
    purgeable_layers: List[int]
    locked_layers: List[int]


def _compute_flux_error(measured: float, reference: float) -> float:
    """Relative flux error: |F̂ - F_ref| / |F_ref|."""
    if abs(reference) < 1e-10:
        return 0.0 if abs(measured) < 1e-10 else float("inf")
    return abs(measured - reference) / abs(reference)


def _compute_centroid_error(
    measured: Tuple[float, float],
    reference: Tuple[float, float],
) -> float:
    """Centroid error: ‖x̂_c - x_{c,ref}‖₂."""
    dx = measured[0] - reference[0]
    dy = measured[1] - reference[1]
    return np.sqrt(dx ** 2 + dy ** 2)


def evaluate_layer(
    decomp: WaveletDecomposition,
    layer: int,
    reference_sources: List[SourceMeasurement],
    epsilon_F: float = 0.005,
    epsilon_x: float = 0.1,
    aperture_radius: float = 5.0,
    annulus_inner: float = 7.0,
    annulus_outer: float = 12.0,
) -> GuardrailResult:
    """
    Evaluate whether a specific residual layer can be safely removed.

    Reconstructs the image without the specified layer (and finer layers),
    then measures all reference sources and checks the science bounds.

    Parameters
    ----------
    decomp : WaveletDecomposition
        Full wavelet decomposition.
    layer : int
        Layer to evaluate for removal (1-indexed).
        Evaluating layer D means: "is reconstruction with
        only layers 1..D-1 (without D and finer) acceptable?"
    reference_sources : list of SourceMeasurement
        Source measurements from the full-fidelity reconstruction.
    epsilon_F : float
        Maximum allowed relative flux error (default 0.5%).
    epsilon_x : float
        Maximum allowed centroid error in pixels (default 0.1).

    Returns
    -------
    GuardrailResult
    """
    # Reconstruct with depth = layer-1 (excluding this layer and finer)
    recon_depth = layer - 1
    reconstructed = reconstruct(decomp, max_depth=recon_depth)

    violations = []
    max_flux_err = 0.0
    max_centroid_err = 0.0

    for ref_src in reference_sources:
        # Measure the same source in the reduced reconstruction
        position = (int(round(ref_src.y_centroid)), int(round(ref_src.x_centroid)))

        # Bounds check
        if (position[0] < 0 or position[0] >= reconstructed.shape[0] or
                position[1] < 0 or position[1] >= reconstructed.shape[1]):
            continue

        meas = measure_source(
            reconstructed, position,
            aperture_radius=aperture_radius,
            annulus_inner=annulus_inner,
            annulus_outer=annulus_outer,
        )

        flux_err = _compute_flux_error(meas.flux, ref_src.flux)
        centroid_err = _compute_centroid_error(
            (meas.x_centroid, meas.y_centroid),
            (ref_src.x_centroid, ref_src.y_centroid),
        )

        max_flux_err = max(max_flux_err, flux_err)
        max_centroid_err = max(max_centroid_err, centroid_err)

        # Check bounds (Eq. 13)
        if flux_err > epsilon_F or centroid_err > epsilon_x:
            violations.append({
                "source_x": ref_src.x_centroid,
                "source_y": ref_src.y_centroid,
                "ref_flux": ref_src.flux,
                "meas_flux": meas.flux,
                "flux_error": flux_err,
                "centroid_error": centroid_err,
            })

    is_safe = len(violations) == 0

    return GuardrailResult(
        layer=layer,
        is_safe=is_safe,
        max_flux_error=max_flux_err,
        max_centroid_error=max_centroid_err,
        num_sources_evaluated=len(reference_sources),
        violations=violations,
    )


def evaluate_frame(
    decomp: WaveletDecomposition,
    reference_sources: List[SourceMeasurement],
    epsilon_F: float = 0.005,
    epsilon_x: float = 0.1,
    aperture_radius: float = 5.0,
    annulus_inner: float = 7.0,
    annulus_outer: float = 12.0,
) -> FrameGuardrailDecision:
    """
    Evaluate all residual layers for a single frame.

    Layers are evaluated in reverse detail order (Section VI.C):
    finest first (H3 → H2 → H1). This allows fine-grained control
    where the outermost layer might be purgeable while inner layers
    must be retained.

    Parameters
    ----------
    decomp : WaveletDecomposition
        Full wavelet decomposition of the frame.
    reference_sources : list of SourceMeasurement
        Source measurements from the full-fidelity reconstruction.
    epsilon_F : float
        Flux error bound.
    epsilon_x : float
        Centroid error bound.

    Returns
    -------
    FrameGuardrailDecision
    """
    layer_decisions = {}
    purgeable = []
    locked = []

    # Evaluate from finest (outermost) to coarsest
    # For 3-level decomposition: evaluate H3, H2, H1
    for layer in range(decomp.levels, 0, -1):
        result = evaluate_layer(
            decomp, layer, reference_sources,
            epsilon_F=epsilon_F,
            epsilon_x=epsilon_x,
            aperture_radius=aperture_radius,
            annulus_inner=annulus_inner,
            annulus_outer=annulus_outer,
        )
        layer_decisions[layer] = result

        if result.is_safe:
            purgeable.append(layer)
        else:
            locked.append(layer)
            # If removing layer D is not safe, all finer layers
            # (D+1, D+2, ...) that were marked purgeable remain so,
            # but we lock everything from D down to 1
            for inner_layer in range(layer - 1, 0, -1):
                if inner_layer not in layer_decisions:
                    # Don't need to evaluate: it's automatically locked
                    layer_decisions[inner_layer] = GuardrailResult(
                        layer=inner_layer,
                        is_safe=False,
                        max_flux_error=float("inf"),
                        max_centroid_error=float("inf"),
                        num_sources_evaluated=0,
                        violations=[{"reason": "locked by parent layer"}],
                    )
                    locked.append(inner_layer)
            break

    return FrameGuardrailDecision(
        frame_id=0,  # Set by caller
        layer_decisions=layer_decisions,
        purgeable_layers=sorted(purgeable, reverse=True),
        locked_layers=sorted(locked),
    )

File: embedded/src/guardrail.cpp
#include "guardrail.hpp"
#include <cmath>
#include <algorithm>
#include <numeric>
#include <limits>

namespace eventvault {
namespace science {

SourceMeasurement measure_source(
    const codec::Matrix2D& image,
    int cx, int cy,
    double aperture_radius,
    double annulus_inner,
    double annulus_outer) 
{
    int rows = image.rows;
    int cols = image.cols;
    
    int y_min = std::max(0, static_cast<int>(cy - annulus_outer - 1));
    int y_max = std::min(rows, static_cast<int>(cy + annulus_outer + 2));
    int x_min = std::max(0, static_cast<int>(cx - annulus_outer - 1));
    int x_max = std::min(cols, static_cast<int>(cx + annulus_outer + 2));
    
    std::vector<double> bg_values;
    
    for (int y = y_min; y < y_max; ++y) {
        for (int x = x_min; x < x_max; ++x) {
            double dy = static_cast<double>(y - cy);
            double dx = static_cast<double>(x - cx);
            double r = std::sqrt(dx*dx + dy*dy);
            
            if (r >= annulus_inner && r <= annulus_outer) {
                bg_values.push_back(image(y, x));
            }
        }
    }
    
    double bg_mean = 0.0;
    double bg_std = 1.0;
    
    if (!bg_values.empty()) {
        std::sort(bg_values.begin(), bg_values.end());
        bg_mean = bg_values[bg_values.size() / 2]; // median
        
        double sum = 0.0, sq_sum = 0.0;
        for (double v : bg_values) {
            sum += v;
            sq_sum += v * v;
        }
        double mean = sum / bg_values.size();
        bg_std = std::sqrt(std::max(0.0, (sq_sum / bg_values.size()) - (mean * mean)));
    }
    
    int aperture_npix = 0;
    double flux = 0.0;
    double peak_value = -1e9;
    double w_sum = 0.0;
    double x_sum = 0.0;
    double y_sum = 0.0;
    
    for (int y = y_min; y < y_max; ++y) {
        for (int x = x_min; x < x_max; ++x) {
            double dy = static_cast<double>(y - cy);
            double dx = static_cast<double>(x - cx);
            double r = std::sqrt(dx*dx + dy*dy);
            
            if (r <= aperture_radius) {
                aperture_npix++;
                double val = image(y, x);
                if (val > peak_value) peak_value = val;
                
                double sub_val = val - bg_mean;
                flux += sub_val;
                
                double weight = std::max(0.0, sub_val);
                w_sum += weight;
                x_sum += weight * static_cast<double>(x);
                y_sum += weight * static_cast<double>(y);
            }
        }
    }
    
    SourceMeasurement meas;
    meas.background_mean = bg_mean;
    meas.aperture_npix = aperture_npix;
    
    if (aperture_npix == 0) {
        meas.x_centroid = static_cast<double>(cx);
        meas.y_centroid = static_cast<double>(cy);
        meas.flux = 0.0;
        meas.snr = 0.0;
        meas.peak_value = 0.0;
        return meas;
    }
    
    meas.flux = flux;
    meas.peak_value = peak_value;
    
    if (flux > 0.0 && w_sum > 0.0) {
        meas.x_centroid = x_sum / w_sum;
        meas.y_centroid = y_sum / w_sum;
    } else {
        meas.x_centroid = static_cast<double>(cx);
        meas.y_centroid = static_cast<double>(cy);
    }
    
    double noise_sq = aperture_npix * (bg_std * bg_std + bg_mean);
    meas.snr = (noise_sq > 0) ? (flux / std::sqrt(noise_sq)) : 0.0;
    
    return meas;
}

std::vector<SourceMeasurement> extract_sources(
    const codec::Matrix2D& image,
    double min_snr,
    double aperture_radius,
    double annulus_inner,
    double annulus_outer,
    double threshold_sigma)
{
    // 1. Estimate background using local mean (approximate uniform filter)
    int box_size = 64;
    int half_box = box_size / 2;
    codec::Matrix2D bg(image.rows, image.cols);
    for (int y = 0; y < image.rows; ++y) {
        for (int x = 0; x < image.cols; ++x) {
            double sum = 0.0;
            int count = 0;
            for (int dy = -half_box; dy <= half_box; ++dy) {
                for (int dx = -half_box; dx <= half_box; ++dx) {
                    int ny = y + dy;
                    int nx = x + dx;
                    if (ny >= 0 && ny < image.rows && nx >= 0 && nx < image.cols) {
                        sum += image(ny, nx);
                        count++;
                    }
                }
            }
            bg(y, x) = sum / count;
        }
    }
    
    codec::Matrix2D sub(image.rows, image.cols);
    std::vector<double> abs_sub;
    abs_sub.reserve(image.rows * image.cols);
    for (size_t i = 0; i < image.data.size(); ++i) {
        double diff = image.data[i] - bg.data[i];
        sub.data[i] = diff;
        abs_sub.push_back(std::abs(diff));
    }
    
    std::sort(abs_sub.begin(), abs_sub.end());
    double mad = abs_sub[abs_sub.size() / 2];
    double noise = mad * 1.4826;
    if (noise <= 0) noise = 1.0;
    
    double threshold = threshold_sigma * noise;
    
    std::vector<std::pair<int, int>> positions;
    int min_sep = 5;
    int half_sep = min_sep / 2;
    
    for (int y = 0; y < image.rows; ++y) {
        for (int x = 0; x < image.cols; ++x) {
            double val = sub(y, x);
            if (val > threshold) {
                bool is_max = true;
                for (int dy = -half_sep; dy <= half_sep && is_max; ++dy) {
                    for (int dx = -half_sep; dx <= half_sep && is_max; ++dx) {
                        if (dy == 0 && dx == 0) continue;
                        int ny = y + dy;
                        int nx = x + dx;
                        if (ny >= 0 && ny < image.rows && nx >= 0 && nx < image.cols) {
                            if (sub(ny, nx) >= val) {
                                is_max = false;
                            }
                        }
                    }
                }
                if (is_max) {
                    positions.push_back({y, x});
                }
            }
        }
    }
    
    std::vector<SourceMeasurement> sources;
    for (const auto& pos : positions) {
        SourceMeasurement meas = measure_source(image, pos.second, pos.first, aperture_radius, annulus_inner, annulus_outer);
        if (meas.snr >= min_snr) {
            sources.push_back(meas);
        }
    }
    
    return sources;
}

static double compute_flux_error(double measured, double reference) {
    if (std::abs(reference) < 1e-10) {
        return (std::abs(measured) < 1e-10) ? 0.0 : std::numeric_limits<double>::infinity();
    }
    return std::abs(measured - reference) / std::abs(reference);
}

static double compute_centroid_error(double mx, double my, double rx, double ry) {
    double dx = mx - rx;
    double dy = my - ry;
    return std::sqrt(dx*dx + dy*dy);
}

GuardrailResult evaluate_layer(
    const codec::WaveletDecomposition& decomp,
    int layer,
    const std::vector<SourceMeasurement>& reference_sources,
    double epsilon_F,
    double epsilon_x,
    double aperture_radius,
    double annulus_inner,
    double annulus_outer)
{
    int recon_depth = layer - 1;
    codec::Matrix2D reconstructed = codec::reconstruct(decomp, recon_depth);
    
    GuardrailResult result;
    result.layer = layer;
    result.max_flux_error = 0.0;
    result.max_centroid_error = 0.0;
    result.num_sources_evaluated = reference_sources.size();
    
    for (const auto& ref_src : reference_sources) {
        int cx = static_cast<int>(std::round(ref_src.x_centroid));
        int cy = static_cast<int>(std::round(ref_src.y_centroid));
        
        if (cx < 0 || cx >= reconstructed.cols || cy < 0 || cy >= reconstructed.rows) {
            continue;
        }
        
        SourceMeasurement meas = measure_source(
            reconstructed, cx, cy, 
            aperture_radius, annulus_inner, annulus_outer
        );
        
        double flux_err = compute_flux_error(meas.flux, ref_src.flux);
        double centroid_err = compute_centroid_error(
            meas.x_centroid, meas.y_centroid, 
            ref_src.x_centroid, ref_src.y_centroid
        );
        
        result.max_flux_error = std::max(result.max_flux_error, flux_err);
        result.max_centroid_error = std::max(result.max_centroid_error, centroid_err);
        
        if (flux_err > epsilon_F || centroid_err > epsilon_x) {
            GuardrailViolation violation;
            violation.source_x = ref_src.x_centroid;
            violation.source_y = ref_src.y_centroid;
            violation.ref_flux = ref_src.flux;
            violation.meas_flux = meas.flux;
            violation.flux_error = flux_err;
            violation.centroid_error = centroid_err;
            violation.reason = "Bounds exceeded";
            result.violations.push_back(violation);
        }
    }
    
    result.is_safe = result.violations.empty();
    return result;
}

FrameGuardrailDecision evaluate_frame(
    const codec::WaveletDecomposition& decomp,
    const std::vector<SourceMeasurement>& reference_sources,
    double epsilon_F,
    double epsilon_x,
    double aperture_radius,
    double annulus_inner,
    double annulus_outer)
{
    FrameGuardrailDecision decision;
    decision.frame_id = 0; // Set by caller
    
    for (int layer = decomp.levels; layer > 0; --layer) {
        GuardrailResult result = evaluate_layer(
            decomp, layer, reference_sources,
            epsilon_F, epsilon_x, aperture_radius, annulus_inner, annulus_outer
        );
        
        decision.layer_decisions[layer] = result;
        
        if (result.is_safe) {
            decision.purgeable_layers.push_back(layer);
        } else {
            decision.locked_layers.push_back(layer);
            
            for (int inner_layer = layer - 1; inner_layer > 0; --inner_layer) {
                if (decision.layer_decisions.find(inner_layer) == decision.layer_decisions.end()) {
                    GuardrailResult auto_locked;
                    auto_locked.layer = inner_layer;
                    auto_locked.is_safe = false;
                    auto_locked.max_flux_error = std::numeric_limits<double>::infinity();
                    auto_locked.max_centroid_error = std::numeric_limits<double>::infinity();
                    auto_locked.num_sources_evaluated = 0;
                    
                    GuardrailViolation v;
                    v.reason = "locked by parent layer";
                    auto_locked.violations.push_back(v);
                    
                    decision.layer_decisions[inner_layer] = auto_locked;
                    decision.locked_layers.push_back(inner_layer);
                }
            }
            break;
        }
    }
    
    std::sort(decision.purgeable_layers.begin(), decision.purgeable_layers.end(), std::greater<int>());
    std::sort(decision.locked_layers.begin(), decision.locked_layers.end());
    
    return decision;
}

} // namespace science
} // namespace eventvault
